### PR TITLE
Suggestions

### PR DESCRIPTION
**Pull request by gyfchong**
Here's a preliminary kind of clean up. I'll continue to make changes over the next day or so because I think there's a good level of complexity here to show you some alternative component architecture to what you've set out.

Of course, feel free to question anything and everything I've done here!

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import React, { Component } from 'react';
```
Comment (gyfchong) [inline src/index.js:1]: I renamed this file to `index` because from the get-go I was confused as to which of your components was the final export. This is also a well known pattern, which doesn't actually impact the name of your component.

File: src/index.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import ListContainer from './ListContainer';
+import ListItemComponent from './ListItemComponent';
+
+const redditAPI = 'https://www.reddit.com/r/';
+
+export default class SubredditPosts extends Component {
```
Comment (gyfchong) [inline src/index.js:7]: I also renamed the exported Class. This is to match your actual component name, and normally consumers would name their import the same name as the package they installed.

I understand that in plain English you are actually getting the sub-reddit from reddit posts, but from the user of this component's view, I think this would be clearer.

File: src/index.js [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    if (extension === 'gif' || extension === 'jpg' || extension === 'jpeg') {
+      return post.data.url;
+    }
+
+    return this.props.placeholder;
+  }
+
+  render() {
```
Comment (gyfchong) [inline src/index.js:52]: Always put your `render` at the bottom of your component. This is not only a recognised pattern, it's also the last thing that your component does in the lifecycle. Also from my perspective every function before the render is basically preparing the component (whether it's fetching data, setting state, computing something, manipulating structure or CSS) for the final result of the component.

File: src/ListContainer.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    case 'listing':
+      return <ul style={flexContainerList}>{children}</ul>;
+
+    default:
+      return <ul style={flexContainerList}>{children}</ul>;
+  }
+}
+
+export default ListContainer;
```
Comment (gyfchong) [inline src/ListContainer.js:44]: I moved this into it's own component because what we want to achieve is "single purpose" components. I gave the main `RedditPosts` (now `index`) component the purpose of fetching the posts and creating the state. Which means that now we can pass the reddit posts into any component we want, it's now agnostic of what the data actually looks like.

The purpose of this `ListContainer` component is to determine the styles and wrap it's children in the appropriate container, and presto you get the desired affect, **no matter the children**. This sets us up for providing more flexibility (which I will be adding tomorrow) in how you would let consumers display the content, ie. Grid view, but using `RedditListing` as the display component.

File: src/ListItemComponent.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import React from 'react';
+import RedditPostIt from './RedditPostIt';
+import RedditListing from './RedditListing';
+import RedditTile from './RedditTile';
+
+const ListItemComponent = ({ display, ...props }) => {
```
Comment (gyfchong) [inline src/ListItemComponent.js:6]: Much like the `ListContainer`, this will provide a more flexible architecture as this will able to decide the display component independently from the `ListContainer`.

File: src/index.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        {this.state.isLoading && <p>Loading...</p>}
+        {!this.state.isLoading && this.state.redditPosts.map(post => (
+          <ListItemComponent
+            display={this.props.display}
+            key={post.key}
+            link={post.link}
+            media={post.media}
+            title={post.title}
+            height={this.props.height}
+            width={this.props.width}
```
Comment (gyfchong) [inline src/index.js:65]: I'll dive into this a bit more in my next set of changes, but I wouldn't let your consumers control the width and height of your components. Your components should always respond to their parent's width to avoid any awkward breakpoints in responsive CSS.

File: src/ListContainer.js [hunk 7]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+
+const flexContainerGallery = {
```
Comment (gyfchong) [inline src/ListContainer.js:3]: Should look into CSS Modules or `emotion` CSS, because these inject the CSS inline which doesn't lend itself to a great developer experience when inspecting and tweaking values to see their effect, can't use pseudo classes, can't use media queries, can't animate... And it's just generally frowned upon.

File: src/index.js [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+  componentDidMount() {
+    const uri = `${redditAPI}${this.props.subreddit}.json`;
+
+    fetch(uri)
+      .then(data => data.json())
+      .then(this.handlePosts)
+      .catch(err => console.error(err));
+  }
+
+  handlePosts = (posts) => {
```
Comment (gyfchong) [inline src/index.js:25]: Oh yeah, the fat arrow functions automatically bind `this`, so you no longer need to define the binds in the constructor.

File: src/index.js (hunk 3 above)
Comment (stcalica) [reply src/index.js:52]: ahh yes I think that was a change left by the last pull request.

File: src/index.js (hunk 1 above)
Comment (stcalica) [reply src/index.js:1]: gotcha, `index.js` means final export file and I don't have to have the same name as the component here! Thanks

File: src/ListContainer.js (hunk 4 above)
Comment (stcalica) [reply src/ListContainer.js:44]: Nice I was trying to find out how to do that. But I was told inheritance is not usually wanted in React. 

I liked the whole `{children}` and `{...props}` attributes. I did not know about them before. 
Any more reading you can point me to about that?

File: src/index.js (hunk 6 above)
Comment (stcalica) [reply src/index.js:65]: no i was wrong about how I wrote that. height and width actually control the size of each image. 
because originally when I wrote this, it was for imgur and in imgur photos are of different sizes and lengths. this created strange and ugly output.

File: src/ListContainer.js (hunk 7 above)
Comment (stcalica) [reply src/ListContainer.js:3]: would those have to be bundled with my dist? 
trying to do as much "vanilla react" as possible with very very little dependence on outside libraries. 
I'm willing to give a look tho if it doesn't damage my requirements.

File: src/index.js (hunk 8 above)
Comment (stcalica) [reply src/index.js:25]: so you changed it from an arrow function and back to: 

Oh nice, would it be recommended to use arrow functions in components to avoid that extra step and that extra LOC?

**Comment by stcalica** [comment]
Awesome thanks for your contribution. I have already reviewed and tested your code, but before I accept please answer a few questions of mine to help me improve my code and my next react project. 

1) Where do you think I can improve? I've been programming for years and my skills need to shape up. I'd think after 6 years I'd code better but I feel like I have a huge lack of skills. I can build things, but not as cleanly as I wished. Any recommendations for readings, practice, etc that you've seen I can improve on now that you've dug through my code? 


2) thanks for the awesome knowledge share! Looking forward to the second pull request!

File: src/ListContainer.js (hunk 4 above)
Comment (gyfchong) [reply src/ListContainer.js:44]: Children are documented a bit in the React docs, unfortunately you have to kind of hit a point where you think "I just want to wrap the components with something and have it spit out the same components" to go searching for something like using `{children}`, but here's an article my Max https://mxstbr.blog/2017/02/react-children-deepdive/

The `{...props}` is something to use carefully, as it's quite ambiguous in exactly what props you're passing, but useful if you know for sure that every prop needs to be passed on. It's simply using the Spread operator on React components: https://davidwalsh.name/spread-operator

File: src/index.js (hunk 6 above)
Comment (gyfchong) [reply src/index.js:65]: Take a look at possibly using `object-fit` in CSS to determine a certain box and force the image to fit into it without compromising the ratio: https://css-tricks.com/almanac/properties/o/object-fit/

Another way is to use CSS Grid and do a Masonry look: https://medium.com/@andybarefoot/a-masonry-style-layout-using-css-grid-8c663d355ebb

File: src/ListContainer.js (hunk 7 above)
Comment (gyfchong) [reply src/ListContainer.js:3]: Yeah, Webpack would create CSS files from `emotion` or CSS Modules, which would be bundled with your dist. But to my knowledge, your consumers won't need to install them as dependencies to use your components, as your components would reference the css files generated at build.

File: src/index.js (hunk 8 above)
Comment (gyfchong) [reply src/index.js:25]: Yeah, fat arrow functions are the recommended way of defining functions which require the use of `this`. However it's not worth putting them on lifecycle methods as it doesn't make a difference and can hurt performance.

File: src/ListContainer.js (hunk 4 above)
Comment (gyfchong) [reply src/ListContainer.js:44]: Oh and I don't know if you've read this yet, but this might be helpful: https://reactjs.org/docs/composition-vs-inheritance.html

I get what you mean, but I don't know enough about programming to be able to explain why this wouldn't be classed as inheritance lol. Maybe the fact that it's not a copy of the parent, it's simply accepting parameters from the parent? Think of it as more of a function.

**Comment by gyfchong** [comment]
> Where do you think I can improve? I've been programming for years and my skills need to shape up. I'd think after 6 years I'd code better but I feel like I have a huge lack of skills. I can build things, but not as cleanly as I wished. Any recommendations for readings, practice, etc that you've seen I can improve on now that you've dug through my code?

Difficult question! To be honest you're already on the right path by doing small projects, and the best way to learn is by doing. I think the way to get better is by writing components for yourself to re-use in bigger projects (ie. you could try build the entirety of Reddit and try use this component), then you'll start to see the challenges in re-usability and how to create better structure to achieve a certain thing. It's a bit difficult outside of working with it everyday, but still a somewhat good simulation.

Also, it's okay to be messy, as long as you revisit every now and then to re-assess and see how you could improve it. My code yesterday is always shit :)

**Comment by gyfchong** [comment]
> thanks for the awesome knowledge share! Looking forward to the second pull request!

Oh and no problem! Always happy to help someone who is eager to learn. I look forward to your next component!

File: src/ListContainer.js (hunk 4 above)
Comment (stcalica) [reply src/ListContainer.js:44]: Just read that last one. Composition vs Inheritance is different cause in composition your using other objects "baked-in" to your "main" object is how I think of it. 
You can use the other pieces for other things or standalone. 
Inheritance it's getting it's underlying structure from it's parent. 
Think about making a cake with ingredients and then the cake can have different forms and functions depending on the event. 

All the ingredients are objects. 
And a wedding cake can inherit from a standard cake, etc.

File: src/index.js (hunk 6 above)
Comment (stcalica) [reply src/index.js:65]: these are interesting. I'll consider this for the imgur react component. 

the grid layout is actually designed for a friend's website and it's exactly how he wants it. 
So I'm not going to try and touch it up too much unfortunately but that's an amazing design solution for the different sizes problem.

File: src/ListContainer.js (hunk 7 above)
Comment (stcalica) [reply src/ListContainer.js:3]: nice i'll def look into it then!

**Comment by stcalica** [comment]
So I actually rewrote it using the methods you showed.  So even though your changes didn't get "pulled" through git. 

I "pulled" them in by hand. It was a good learning process to refactor. Thank you!

File: src/RedditListing.js [hunk 9]
```diff
@@ -22,12 +22,10 @@ export default function RedditListing(props) {
...
-  const listTitle = {
+  const titleListItem = {
     margin: '10px'
   };
 
-  const titleListItem = {};
-
   return (
     <li style={listItem}>
       <a style={anchorStyle} href={props.link}>
```
Comment (gyfchong) [inline src/RedditListing.js:31]: So I was trying to do my second PR, but realised it was a bit of refactoring to get it working and you should give it a go yourself! Basically since all your RedditPost-X were the same HTML, the best practice would be to abstract them into their own component.

Comment (stcalica) [reply src/RedditListing.js:31]: See that's where I was thinking of doing inheritance but fell into composition based off the docs. 
What do you mean by abstraction when it comes to react? A single component and the passed in props would select the CSS, that was my second guess. 

Yeah I was able to get to work. I'll be pushing my new code and publishing later today. It's pretty much exactly what you have here.

Comment (gyfchong) [reply src/RedditListing.js:31]: Ah I see what you mean now. But I think in general when people in React talk about "avoiding inheritance" they're mainly talking about extending the class `React.Component`, which provides all the lifecycle methods and state.

Theoretically there's nothing wrong with creating every component that inherits from `React.Component`, practically it might have some perf issues, but we try to avoid that in most cases because lifting state makes it easier to reason about your application's state, attempting to keep it all in one place (https://reactjs.org/docs/lifting-state-up.html) and trying to create more functional components clearly outlines the components which exist only to provide visuals versus those that hold logic (https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0). And you'll start to see why this pattern is desirable when you start making bigger applications as components need to be re-used for their visual aspect more often than components that deal with data. Maybe check out my Hcard Builder for a clearer example of this at a small-medium scale: https://github.com/gyfchong/hcard-builder

Abstraction in this context just means move the code into it's own component so you can re-use it in different places.

... Hope that helps!

Comment (gyfchong) [reply src/RedditListing.js:31]: > A single component and the passed in props would select the CSS, that was my second guess.

Yeah that's the right direction. I would look into implementing CSS modules first, so you can take advantage of conditional classnames (https://github.com/JedWatson/classnames) which will help clean up your code.

Comment (stcalica) [reply src/RedditListing.js:31]: Awesome thanks so much! 
Def love all the links!
